kano_updater/monitor.py

```diff
@@ -0,0 +1,165 @@
+# monitor.py
+#
+# Copyright (C) 2014-2018 Kano Computing Ltd.
+# License: http://www.gnu.org/licenses/gpl-2.0.txt GNU GPL v2
+#
+# Module to allow us to detect when a process is stuck.
+
+# Also heartbeat() function for a monitored process
+# to notify the monitor that it is proceeding normally.
+
+import subprocess
+from collections import defaultdict
+import sys
+import time
+import signal
+import os
+from kano.logging import logger
+import kano_updater.utils
+
+MONITOR_TIMEOUT = 20 * 60
+TIMEOUT_RC = 105
+
+
+class monitorPids:
+    """
+    Class for monitoring a subprocess tree. If the (recursive) set of child
+    processes changes, we assume it is still making progress.
+    """
+    def __init__(self, top_pid):
+        self.top_pid = top_pid
+        self.curr_children = set()
+
+    def _get_children(self):
+        """
+        return a set() of all children (recursively) of
+        self.top_pid (inclusive)
+        """
+        try:
+            stdout = subprocess.check_output(["/bin/ps", "-eo", "ppid,pid"])
+            coll = defaultdict(set)
+
+            for line in stdout.split('\n'):
+                items = line.split()
+                if len(items) >= 2:
+                    ppid, pid = items[:2]
+                    if ppid == 'PPID':
+                        continue
+            try:
+                coll[int(ppid)].add(int(pid))
+            except:
+                pass  # ignore int conversion failure
+
+            def closure(pids):
+                curr = pids.copy()
+                for p in pids:
+                    curr = curr.union(closure(coll[p]))
+                return curr
+            return closure(set([self.top_pid]))
+
+        except subprocess.CalledProcessError:
+            return set()
+
+    def isChanged(self):
+        """
+        return True if we believe it is making progress
+        otherwise False
+        """
+        changed = False
+        new_children = self._get_children()
+
+        if self.curr_children != new_children:
+            self.curr_children = new_children
+            changed = True
+        return changed
+
+
+def monitor(watchproc, timeout):
+    """
+    Monitor a process tree. If the (recursive) set of child processes changes
+    or we are sent a SIGUSR1, we note that it is making progress.
+    if it has not made progress for `timeout` seconds, or the process
+    finished, exit.
+
+    Returns true if we timed out and false if the process finished
+
```

Comment:
Arguments in either [this form](http://dev.kano.me/docs/kano-toolset/_modules/kano/utils/file_operations.html#touch) so sphinx can parse them.